base/array.jl

```diff
@@ -1741,6 +1741,13 @@ end
 findfirst(testf::Function, A::Union{AbstractArray, AbstractString}) =
     findnext(testf, A, first(keys(A)))
 
+function findfirst(p::Union{Fix2{typeof(isequal),T},Fix2{typeof(==),T}}, r::StepRange{T,S}) where {T,S}
```

Comment:
This should perhaps be restricted to Integer types. It's tricky to know whether this formula will hit a floating point value exactly; see for example the `_in_range` function in this file. You have to check that the value in the range actually equals what you're looking for. Restricting to Integers also makes it easy to allow different types; the signature can be
```
findfirst(p::Union{Fix2{typeof(isequal),<:Integer},Fix2{typeof(==),<:Integer}}, r::StepRange{<:Integer,S}) where {S}
```

Reply:
My actual use case was with `DateTime` objects, for example:
```julia-repl
julia> @btime findfirst(==(DateTime(2019, 1, 23)), DateTime(2017):Second(1):DateTime(2020)) # before
  53.385 ms (3 allocations: 64 bytes)
64972801

julia> @btime findfirst(==(DateTime(2019, 1, 23)), DateTime(2017):Second(1):DateTime(2020)) # after
  64.659 ns (0 allocations: 0 bytes)
64972801
```

The fact that `T` occurs in `StepRange{T,S}` should restrict it to discrete types.

Reply:
Good point. It's too bad that the method won't match for different types of integers, but I guess this is ok for now.